### PR TITLE
Add a skeleton for repetitively running tests

### PR DESCRIPTION
**Pull request by niyue**

### CODE REVIEW

File: tests/topology_test.cpp [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        std::this_thread::sleep_for(std::chrono::milliseconds(milliseconds));
+    }
+}
+
+TEST_CASE_METHOD(VectorService<stdin_to_file>, "start stdin to file topology") {
+   _send_events_to_stdin("events.log");
+   _wait();
+   auto events = _read_file_by_lines(FILE_SINK_PATH);
+   REQUIRE(events.size() == 10);
+}
```
Comment (niyue) [inline tests/topology_test.cpp:123]: This test case can be run repetitively and assert the results of the topology. We could follow this to add more tests.

File: tests/topology_test.cpp [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+// TEST_CASE_METHOD(VectorService<stdin_to_file_with_transform>, "start topology with transformation") {
+//     _send_events_to_stdin("events.log");
+//     _wait();
+//     auto events = _read_file_by_lines(FILE_SINK_PATH);
+//     REQUIRE(events.size() == 9);
+// }
```
Comment (niyue) [inline tests/topology_test.cpp]: I would like to add some tests with transformation, but apparently, the second tests is affected by the first test case, and the second test case always produce 0 zero results when running after 1st test case. I will do some more investigation later to see what causes this issue.

File: .clang-format [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+Language: Proto
+# Don't format .proto files.
+DisableFormat: true
+---
+Language: ObjC
+# clang-format misunderstands some header files as Objective-C files
+DisableFormat: true
+---
+Language: Java
+ColumnLimit:     100
```
Comment (niyue) [inline .clang-format:18]: Add a clang-format for formatting C++

File: src/lib.rs [hunk 4]
```diff
@@ -1,9 +1,19 @@
...
+
 use tokio_test::block_on;
 use std::panic;
 use std::sync::{Arc, Mutex};
 use crate::ffi::{ExportResult, KafkaSinkParams, SwEvent, SwEvents};
+use std::sync::Once;
+use std::fmt::{Debug, Display};
+
 // use futures::executor::block_on;
+use tracing::{debug, info, error, Level};
```
Comment (niyue) [inline src/lib.rs:11]: Use the `tracing` library for logging

File: src/lib.rs [hunk 5]
```diff
@@ -95,27 +105,50 @@ enum ConfigAction {
...
 
+pub fn setup_logging() {
+    let timer = tracing_subscriber::fmt::time::time();
+    let collector = tracing_subscriber::fmt()
+        .with_max_level(Level::DEBUG)
+        // disable color to make CLion happy
+        .with_ansi(false)
+        .with_thread_ids(true)
+        .with_timer(timer)
+        .finish();
```
Comment (niyue) [inline src/lib.rs:126]: Here are the settings for logging library, currently, debug level log is enabled.

File: tests/main_test.cpp [hunk 6]
```diff
@@ -1,275 +0,0 @@
-#include "vectorcxx/src/lib.rs.h"
```
Comment (niyue) [inline tests/main_test.cpp:1]: I delete the entire test since it is covered by the topology_test now. We should make all the tests repeatable

File: tests/topology_test.cpp [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+  std::vector<std::string>
+  _read_events_from_sink(const std::filesystem::path &file_path = FILE_SINK_PATH) {
+    return _read_file_by_lines(file_path);
+  }
+
+  void _wait(uint32_t milliseconds = 200) {
+    std::this_thread::sleep_for(std::chrono::milliseconds(milliseconds));
+  }
+
+  void send_http_events(const std::vector<std::string> &events, uint32_t port = 9999) {
```
Comment (niyue) [inline tests/topology_test.cpp:57]: I switch to use http source for testing since it is simpler than problematic stdin

File: CMakeLists.txt [hunk 8]
```diff
@@ -1,16 +1,17 @@
...
+message(STATUS "#################### PROJECT ${PROJECT_NAME} ####################") 
+
 # https://vcpkg.readthedocs.io/en/latest/users/integration/#cmake
 if(DEFINED ENV{VCPKG_ROOT}) 
     message(STATUS "vcpkg found VCPKG_ROOT=$ENV{VCPKG_ROOT}")
     # https://github.com/microsoft/vcpkg/blob/master/docs/users/integration.md#changing-the-triplet-1
 
     if(DEFINED ENV{VCPKG_DEFAULT_TRIPLET})
-        message(STATUS "using VC_DEFAULT_TRIPLET from environment $ENV{VCPKG_DEFAULT_TRIPLET}")
+        message(STATUS "using VCPKG_DEFAULT_TRIPLET from environment $ENV{VCPKG_DEFAULT_TRIPLET}")
```
Comment (ProBrian) [inline CMakeLists.txt:14]: There are multiple changes for building related stuffs: ENVs, presets, etc. Do we have any update on the README? Or these changes are transparent to developers so we don't need to do any change in development?

Comment (niyue) [reply CMakeLists.txt:14]: It doesn't matter. I generate the Cmake file with latest piemaker and update some dependencies

File: src/lib.rs (hunk 5 above)
Comment (ProBrian) [reply src/lib.rs:126]: Can we make log level configurable, like vector's : "std::env::var("VECTOR_LOG")", so that for debug/release version, we can collect logs for different levels, no need to manually change the code.

File: src/lib.rs [hunk 9]
```diff
@@ -95,27 +102,44 @@ enum ConfigAction {
...
+    tracing::subscriber::set_global_default(collector).expect("setting default subscriber failed");
```
Comment (ProBrian) [inline src/lib.rs:127]: What's the impact of `set_global_default` here? I see there are same usage in vector's code for something like metrics collection as below:
```
        let subscriber = tracing_subscriber::Registry::default().with(MetricsLayer::new());
        tracing::subscriber::set_global_default(subscriber).unwrap();
```
I'm not sure if this will block vector's original functional logic.
As I see in this method's doc(https://docs.rs/tracing/0.1.37/tracing/subscriber/fn.set_global_default.html):
```Can only be set once; subsequent attempts to set the global default will fail. Returns whether the initialization was successful.```. We should make sure not break something else.

Comment (niyue) [reply src/lib.rs:127]: https://github.com/vectordotdev/vector/search?q=set_global_default it is only used in test/benches of vector, so it should be fine
